util/log: de-flake TestSecondaryGC and some SQL tests

knz opened the pull request:
Fixes #57442
Fixes #57444 

See individual commits for details.

knz committed:
util/log: de-flake TestSecondaryGC

The GC daemon was performing an unconditional round of deletion,
asynchronously with the start of the test. This was causing a race
condition.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/57454)
<!-- Reviewable:end -->

knz committed:
util/log: de-flake certain SQL tests

As discussed previously in #57222 - the SQL servers do not obey the
server stop logic well and leak goroutines that continue to log while
the TestLogScope gets closed.

This commit continues to tolerate this misdesign by ensuring that
some more stderr-related flags are accessed atomically.

Release note: None

knz commented:
thanks!

bors r=irfansharif
